desc/integrals/singularities.py

```diff
@@ -327,104 +446,110 @@ def __call__(self, f, i):
+        axisymmetric configurations, we set a dummy value for NFP to an integer larger
+        than 1 so that the toroidal increment can move to a new spot.
+        """
         f, source_data = f_data
-        source_data["zeta"] = (source_zeta + j * 2 * np.pi / source_grid.NFP) % (
-            2 * np.pi
+        source_data["zeta"] = (source_zeta + j * 2 * jnp.pi / source_grid.NFP) % (
+            2 * jnp.pi
         )
-        source_data["phi"] = (source_phi + j * 2 * np.pi / source_grid.NFP) % (
-            2 * np.pi
+        source_data["phi"] = (source_phi + j * 2 * jnp.pi / source_grid.NFP) % (
+            2 * jnp.pi
         )
 
         # nest this def to avoid having to pass the modified source_data around the loop
         # easier to just close over it and let JAX figure it out
-        def eval_pt_vmap(i):
-            # this calculates the effect at a single evaluation point, from all others
-            # in a single field period. vmap this to get all pts
-            k = kernel(
-                {key: val[i] for key, val in eval_data.items() if key in keys},
-                source_data,
-            ).reshape((1, source_grid.num_nodes, kernel.ndim))
-
-            rho = _rho(
+        def eval_pt(eval_data_i):
+            k = kernel(eval_data_i, source_data).reshape(
+                -1, source_grid.num_nodes, kernel.ndim
+            )
+            eta = _eta(
                 source_theta,
-                source_data["zeta"],  # to account for different field periods
-                eval_theta[i],
-                eval_zeta[i],
-                h_t,
-                h_z,
-                s,
+                source_data["zeta"],
+                eval_data_i["theta"][:, jnp.newaxis],
+                eval_data_i["zeta"][:, jnp.newaxis],
+                ht,
+                hz,
+                st,
+                sz,
             )
+            return jnp.sum(k * (w * (1 - eta))[..., jnp.newaxis], axis=1)
 
-            eta = _chi(rho)  # from eq 36 of [2]
-            k = (1 - eta)[None, :, None] * k
-            f_temp = jnp.sum(k * w[None, :, None], axis=1)
-            return f_temp
-
-        def eval_pt_loop(i, fj):
-            # this calculates the effect at a single evaluation point, from all others
-            # in a single field period. loop this to get all pts
-            f_temp = eval_pt_vmap(i)
-            return put(fj, i, f_temp.reshape(fj[i].shape))
-
-        # vmap for inner part found more efficient than fori_loop, especially on gpu,
-        # but for jacobian looped seems to be better and less memory
-        if loop:
-            fj = fori_loop(0, eval_grid.num_nodes, eval_pt_loop, jnp.zeros_like(f))
-        else:
-            fj = vmap(eval_pt_vmap)(jnp.arange(eval_grid.num_nodes))
-
-        f += fj.reshape((eval_grid.num_nodes, kernel.ndim))
+        f += batch_map(eval_pt, eval_data, 1 if loop else None).reshape(
```

Comment:
I think we should allow arbitrary batch sizes while we're at it. Just have loop default to batch size=1